rust/src/api.rs

```diff
@@ -135,18 +136,50 @@ fn get_exist_proof<'a>(
     }
 }
 
-fn get_nonexist_proof<'a>(
+fn get_nonexist_proof<'a, H: HostFunctionsProvider>(
     proof: &'a ics23::CommitmentProof,
     key: &[u8],
+    spec: &ics23::ProofSpec,
 ) -> Option<&'a ics23::NonExistenceProof> {
+    // If we should prehash the key before comparison, do so; otherwise, return the key. Prehashing
+    // changes lexical comparison, so if the spec sets `prehash_key_before_comparison` to true, we
+    // should compare keys on their hashes by prehashing them before comparison.
+    let key_for_comparison = |key: &[u8]| match spec.prehash_key_before_comparison {
+        true => do_hash::<H>(
+            spec.leaf_spec.clone().unwrap_or_default().prehash_key(),
```

Comment:
nit: We could avoid cloning here by doing something like this, but it does not read as nice as the current version imho, which makes it clear we are using the default/empty `LeafSpec`. 

```suggestion
            spec.leaf_spec.as_ref().map(|leaf_spec| leaf_spec.prehash_key()).unwrap_or_default(),
            
```